### PR TITLE
[WIP]add npu support

### LINKED ISSUE #19498: Add Ascend NPU as a backend
### Description & Motivation

Ascend is a full-stack AI computing infrastructure for industry applications and services based on Huawei Ascend processors and software. For more information about Ascend, see [Ascend Community](https://www.hiascend.com/en/).

[CANN](https://www.hiascend.com/en/software/cann) (Compute Architecture of Neural Networks), developped by Huawei, is a heterogeneous computing architecture for AI.

Pytorch has officially announced [support for Ascend NPU](https://pytorch.org/blog/pytorch-2-1/) (through key PrivateUse1), please see the PrivateUse1 tutorial [here](https://pytorch.org/tutorials/advanced/privateuseone.html).

Provide new backend support for pytorch-lighting, allowing users who use Ascend NPU to also use the convenient development and acceleration capabilities provided by pytorch-lighting.

### Pitch

I'd like to add a new accelerator, and register it into Accelerators, which make it possible that customer can use this new backend by: "accelerator='npu'".

### Alternatives

_No response_

### Additional context

I have wrote a demo, please refer to: https://github.com/Lightning-AI/pytorch-lightning/pull/19308

cc @borda

### PR DESCRIPTION
**Pull request by hipudding**
## Add Ascend NPU as a new backend for pytorch-lighting

Fixes #19498

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
📚 Documentation preview 📚: https://pytorch-lightning--19308.org.readthedocs.build/en/19308/

### CODE REVIEW

**Comment by hipudding** [comment]
Reference implementation #17700 #19443

**Comment by hhllxx1121** [comment]
Hello, If possible ，can you give me a demo of how to use Lightning on an NPU?

**Comment by hhllxx1121** [comment]
Hello, if possible, could you show me a demo of how to use Lightning on an NPU?

**Comment by GuWei007** [comment]
@hipudding @hhllxx1121 
More and more device manufacturers need to support their own backends. Can this be achieved by using the PrivateUse1 mechanism?

**Comment by hipudding** [comment]
> Hello, if possible, could you show me a demo of how to use Lightning on an NPU?

Yes, it's very simple to use npu instead of other backends. Just replace accelerator to 'npu'. Note that this PR is under development and not pass all the test. DO NOT use for production.

Here's official example with NPU.
```Python
# main.py
# ! pip install torchvision
import torch, torch.nn as nn, torch.utils.data as data, torchvision as tv, torch.nn.functional as F
import lightning as L

# --------------------------------
# Step 1: Define a LightningModule
# --------------------------------
# A LightningModule (nn.Module subclass) defines a full *system*
# (ie: an LLM, diffusion model, autoencoder, or simple image classifier).


class LitAutoEncoder(L.LightningModule):
    def __init__(self):
        super().__init__()
        self.encoder = nn.Sequential(nn.Linear(28 * 28, 128), nn.ReLU(), nn.Linear(128, 3))
        self.decoder = nn.Sequential(nn.Linear(3, 128), nn.ReLU(), nn.Linear(128, 28 * 28))

    def forward(self, x):
        # in lightning, forward defines the prediction/inference actions
        embedding = self.encoder(x)
        return embedding

    def training_step(self, batch, batch_idx):
        # training_step defines the train loop. It is independent of forward
        x, y = batch
        x = x.view(x.size(0), -1)
        z = self.encoder(x)
        x_hat = self.decoder(z)
        loss = F.mse_loss(x_hat, x)
        self.log("train_loss", loss)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=1e-3)
        return optimizer


# -------------------
# Step 2: Define data
# -------------------
dataset = tv.datasets.MNIST(".", download=True, transform=tv.transforms.ToTensor())
train, val = data.random_split(dataset, [55000, 5000])

# -------------------
# Step 3: Train
# -------------------
autoencoder = LitAutoEncoder()
trainer = L.Trainer(accelerator='npu', devices='0,1', max_epochs=1, strategy='deepspeed')
trainer.fit(autoencoder, data.DataLoader(train), data.DataLoader(val))
```

**Comment by hipudding** [comment]
> @hipudding @hhllxx1121 More and more device manufacturers need to support their own backends. Can this be achieved by using the PrivateUse1 mechanism?

As far as I know. pytorch-lightning use abstract accelerator, and not support plugins. If not all backends' accelerator can be merge. accelerator plugin is a good choice.

**Comment by uniartisan** [comment]
> > @hipudding @hhllxx1121 More and more device manufacturers need to support their own backends. Can this be achieved by using the PrivateUse1 mechanism?
> 
> As far as I know. pytorch-lightning use abstract accelerator, and not support plugins. If not all backends' accelerator can be merge. accelerator plugin is a good choice.

https://github.com/Lightning-AI/pytorch-lightning/pull/20349

This pr enhanced plugin experience :)
